src/mcts/search.cc

```diff
@@ -1281,18 +1277,44 @@ void SearchWorker::DoBackupUpdateSingleNode(
     return;
   }
 
+  // For the first visit to a terminal, maybe convert ancestors to terminal too.
+  auto can_convert = params_.GetTerminator() && node->IsTerminal() && !node->GetN();
+
   // Backup V value up to a root. After 1 visit, V = Q.
   float v = node_to_process.v;
   float d = node_to_process.d;
-  for (Node* n = node; n != search_->root_node_->GetParent();
-       n = n->GetParent()) {
+  for (Node *n = node, *p; n != search_->root_node_->GetParent(); n = p) {
+    p = n->GetParent();
+
+    // Current node might have become terminal from some other descendant, so
+    // backup the rest of the way with more accurate values.
+    if (n->IsTerminal()) {
+      v = n->GetQ();
+      d = n->GetD();
+    }
     n->FinalizeScoreUpdate(v, d, node_to_process.multivisit);
+
+    // Convert parents to terminals except the root or those already converted.
+    can_convert = can_convert && p != search_->root_node_ && !p->IsTerminal();
+
+    // A non-winning terminal move needs all other moves to have the same value.
+    if (can_convert && v != 1.0f)
+      for (const auto& edge : p->Edges())
```

Comment:
Braces around body of `for` because it's on a separate line.
Consider this (although I'm not sure that it's more readable):
```cpp
auto edges = p->Edges();
can_convert = std::all_of(edges.begin(), edges.end(), [&](const Edge& e) {
  return e.IsTerminal() && e.GetQ(0.0f) == v;
});
```

Reply:
Switched to `all_of` but with `const auto &` instead of `const Edge&` as it's actually a `EdgeAndNode`.